### PR TITLE
Fix sparse distillation loss and speed up teacher top-k logit saving

### PR DESCRIPTION
**Pull request by ajkv-google**
# Description

This PR fixes a mathematical discrepancy in the offline sparse distillation loss and optimizes the background writing process for top-K teacher logits.

**Problem being solved & relevant context:**
1. During offline sparse distillation, there was a mathematical flaw in how the KL divergence was computed. The teacher's probabilities were normalized strictly over the top-K subset, while the student's log-probabilities were computed over the full vocabulary and then gathered. This led to comparing two distributions with completely different normalization scales.
2. The `save_top_k_teacher_logits` script was experiencing bottlenecks during the CPU-bound serialization step, which could cause the TPU to wait for files to be written to the gs bucket.

**Solution & Implementation:**
- **Sparse KL Divergence Fix:** In `distillation_utils.py`, the student's unnormalized logits are now gathered at the teacher's exact Top-K indices *before* normalization. Then, `log_softmax` is applied exclusively to this subset. This ensures both distributions are normalized correctly. 
- **Saving Logits Optimization:** In `save_top_k_teacher_logits.py`, the single-threaded CPU conversion is now parallelized using a new `serialization_executor`. This maps the `create_tf_example` calls across multiple CPU cores before sequentially writing the serialized records to disk. 

# Tests

Used the following to run offline distillation:

- Command: https://paste.googleplex.com/4852897526448128
- Yaml: https://paste.googleplex.com/4981440528908288
- Example run for offline distillation (using our team's LTI checkpoint for 30b model): [Tensorboard Link](https://pantheon.corp.google.com/storage/browser/ajkv-distillation/offline-teacher-logits-test-run/test-off-dist/qwen3-30b-a3b-base_2026-06-03-20-59/tensorboard?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev)

Used the following to run saving of top-k teacher logits:

- Command: https://paste.googleplex.com/4974440839512064
- Yaml: https://paste.googleplex.com/6389552133570560
- GS Bucket Path for saved logits (1B tokens of climbmix): [Saved offline logits](https://pantheon.corp.google.com/storage/browser/ajkv-distillation/offline-teacher-logits-test-run/test-sav-log?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py [hunk 1]
```diff
@@ -92,26 +89,36 @@ def create_tf_example(example_dict):
...
 
-    flat_val = np.asarray(val).flatten()
+    # Use .ravel() to avoid the memory copy that .flatten() does
+    flat_val = np.asarray(val).ravel()
 
     if flat_val.dtype in [np.float32, np.float64, np.float16, jnp.bfloat16]:
-      features[key] = tf.train.Feature(float_list=tf.train.FloatList(value=flat_val.astype(np.float32)))
+      if flat_val.dtype != np.float32:
+        flat_val = flat_val.astype(np.float32)
+      # Use .tolist() for fast Protobuf C++ ingestion
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py]: remove comment?

File: src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py [hunk 2]
```diff
@@ -163,12 +176,14 @@ def generate_and_save_data(config, local_args):
...
-  # all hosts initialize their own directories and thread pools
   if not os.path.exists(local_tmp_dir):
     os.makedirs(local_tmp_dir, exist_ok=True)
 
   upload_executor = ThreadPoolExecutor(max_workers=4)
-  write_executor = ThreadPoolExecutor(max_workers=2)
+  # Restrict to 1 worker to ensure sequential writing to the ArrayRecord file
+  write_executor = ThreadPoolExecutor(max_workers=1)
+  # New executor purely for CPU-bound protobuf serialization
+  serialization_executor = ThreadPoolExecutor(max_workers=16)
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py]: hardcoded max_workers

File: src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py [hunk 3]
```diff
@@ -204,15 +218,18 @@ def generate_and_save_data(config, local_args):
             if jax.process_index() == 0:
               max_logging.log(f"Queueing distributed background uploads for Step {step}...")
             upload_executor.submit(background_upload, local_output_path, gcs_file_path, jax.process_index())
-          write_executor = ThreadPoolExecutor(max_workers=2)
+
+          # Re-initialize the writer with 1 worker
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py]: it is better to provide some short comment on why we have to do that

**Review by vlad-karp — APPROVED** [review]

File: src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py (hunk 2 above)
Comment (ajkv-google) [reply src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py]: Added them are extra arguments

**Review by JamesDeng42 — APPROVED** [review]
